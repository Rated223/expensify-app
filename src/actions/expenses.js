import database from '../firebase/firebase';

const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
  expense
});

const startAddExpense = (expenseData = {}) => {
  return (dispatch) => {
    const {
      description = '', 
      note = '', 
      amount = 0, 
      createdAt = 0 
    } = expenseData;

    const expense = { description, note, amount, createdAt };
    
    return database.ref('expenses').push(expense)
    .then((ref) => {
      dispatch(addExpense({
        id: ref.key,
        ...expense
      }))
    })
    .catch((err) => {
      console.error('firebase ', err);
    })
  }
}

const removeExpense = ({ id } = {}) => ({
    type: 'REMOVE_EXPENSE',
    id
});

const startRemoveExpense = ({ id }) => {
  return (dispatch) => {
    return database.ref(`expenses/${id}`).remove()
    .then(() => {
      dispatch(removeExpense({ id }));
    })
    .catch((err) => {
      console.error('firebase ', err);
    })
  }
}

const editExpense = ({ id, update }) => ({
  type: 'EDIT_EXPENSE',
  id,
  update
});

const startEditExpense = ({  id, update }) => {
  return (dispatch) => {
    return database.ref(`expenses/${id}`).update({...update})
    .then(() => {
      dispatch(editExpense({ id, update }));
    }).catch((err) => {
      console.error('firebase ', err);
    })
  }
}

const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses
});

const startSetExpenses = () => {
  return (dispatch) => {
    return database.ref('expenses').once('value')
    .then((snapshot) => {
      const expenses = [];

      snapshot.forEach((child) => {
        expenses.push({
          id: child.key,
          ...child.val()
        });
      });

      dispatch(setExpenses(expenses));
    })
    .catch((err) => {
      console.error('firebase ', err);
    })
  }
}

export {
  addExpense,
  startAddExpense,
  removeExpense,
  startRemoveExpense,
  editExpense,
  startEditExpense,
  setExpenses,
  startSetExpenses
}