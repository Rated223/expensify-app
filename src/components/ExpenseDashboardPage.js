import React from 'react';
import { ExpenseList } from '.';

const ExpenseDashboardPage = () => (
  <div>
    <ExpenseList />
  </div>
);

export default ExpenseDashboardPage;