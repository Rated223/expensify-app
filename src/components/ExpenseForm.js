import 'react-dates/initialize';
import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
import 'react-dates/lib/css/_datepicker.css';

const now = moment();
console.log('now', now.format('MMM Do YYYY'));

class ExpenseForm extends React.Component {
  state = {
    description: '',
    amount: '',
    note: '',
    createdAt: moment(),
    calendarFocused: false,
    error: ''
  }

  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  }

  onNoteChange =  (e) => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  }
  
  onAmountChange = (e) => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  }

  onDateChange = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({ createdAt }))
    }
  }

  onCalendarFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }))
  }

  onSubmit = (e) => {
    e.preventDefault();
    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({ error: 'Please provide description and amount' }));
    } else {
      this.setState(() => ({ error: '' }));
      this.props.onSubmit({
        description: this.state.description,
        amount: +this.state.amount,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note

      })
    }
  }

  render() {
    return (
      <div>
        {this.state.error && <p>{this.state.error}</p>}
        <form onSubmit={this.onSubmit}>
          <input
            type="text"
            placeholder="Description"
            value={this.state.description}
            onChange={this.onDescriptionChange}
            autoFocus
          />
          <input
            type="text"
            placeholder="Amount"
            value={this.state.amount} 
            onChange={this.onAmountChange}
          />
          <SingleDatePicker 
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.onCalendarFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <textarea
            placeholder="Add a note for yor expense (optional)"
            value={this.state.notes}
            onChange={this.onNoteChange}
          ></textarea>
          <button>Add expense</button>
        </form>
      </div>
    )
  }
} 

export default ExpenseForm;