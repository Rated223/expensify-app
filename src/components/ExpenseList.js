import React from "react";
import { connect } from "react-redux";
import { ExpenseListItem } from ".";
import { getVisibleExpenses } from '../selectors';

export const ExpenseList = (props) => (
  <div className="content-container">
    <div className="list-header">
      <div className="show-for-mobile">Expenses</div>
      <div className="show-for-desktop">Expense</div>
      <div className="show-for-desktop">Amount</div>
    </div>
    <div className="list-body">
      {
        props.expenses.length === 0 ? (
          <div className="list-item--message">
            <p>No expenses</p>
          </div>
        ) : (
          <div>
            {props.expenses.map((expense) => {
              return <ExpenseListItem key={expense.id} {...expense} />;
            })}
          </div>
        )
      }
    </div>
  </div>
);

const mapStateToProps = (state) => {
  return {
    expenses: getVisibleExpenses(state.expenses, state.filters)
  }
};

export default connect(mapStateToProps)(ExpenseList);