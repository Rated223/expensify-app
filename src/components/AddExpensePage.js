import React from 'react';
import { connect } from 'react-redux';
import { ExpenseForm } from '.';
import { expensesActions } from '../actions';

export class AddExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.addExpense(expense);
    this.props.history.push('/');
  }

  render() {
    return (
      <div>
        <p>Add expense</p>
        <ExpenseForm
          onSubmit={this.onSubmit}
        />
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch) => ({
  addExpense: (expense) => dispatch(expensesActions.addExpense(expense))
})

export default connect(undefined, mapDispatchToProps)(AddExpensePage);
