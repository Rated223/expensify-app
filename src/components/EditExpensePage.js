import React from 'react';
import { connect } from 'react-redux';
import { ExpenseForm } from ".";
import { expensesActions } from "../actions";

export class EditExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.startEditExpense({ id: this.props.expense.id, expense })
    this.props.history.push('/');
  }

  onClick = () => {
    this.props.startRemoveExpense({ id: this.props.expense.id })
    this.props.history.push('/');
  }

  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Edit expense</h1>
          </div>
        </div>
        <div className="content-container">
          <div className="content-container__left">
            <button className="button--secondary" onClick={this.onClick}>
              Remove expense
            </button>
          </div>
          <ExpenseForm 
            expense={this.props.expense}
            onSubmit={this.onSubmit}
          />
        </div>
      </div>
    )
  }
}

const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find((expense) => expense.id === props.match.params.id)
  }
}

const mapDispatchToProps = (dispatch) => ({
  startEditExpense: ({ id, expense }) => dispatch(expensesActions.startEditExpense({ id, update: expense })),
  startRemoveExpense: ({ id }) => dispatch(expensesActions.startRemoveExpense({ id }))
})

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
