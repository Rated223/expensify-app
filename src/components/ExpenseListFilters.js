import 'react-dates/initialize';
import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from "react-dates";
import { filtersActions } from '../actions';
import 'react-dates/lib/css/_datepicker.css';

export class ExpenseListFilters extends React.Component {

  state = {
    calendarFocused: null
  }

  onTextChange = (e) => {
    this.props.setTextFilter({ text: e.target.value });
  }

  onSortChange = (e) => {
    if (e.target.value === "date") {
      this.props.setSortByDate();
    } else if (e.target.value === "amount") {
      this.props.setSortByAmount();
    }
  }
  
  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate({ startDate });
    this.props.setEndDate({ endDate });
  }

  onCalendarFocusChange = (calendarFocused) => {
    this.setState(() => ({ calendarFocused }))
  }

  render() {
    return (
      <div>
        <input 
          type="text" 
          value={this.props.filters.text} 
          onChange={this.onTextChange}
        />
        <select 
          value={this.props.filters.sortBy} 
          onChange={this.onSortChange}
        >
          <option value="date">Date</option>
          <option value="amount">Amount</option>
        </select>
        <DateRangePicker 
          startDate={this.props.filters.startDate}
          startDateId="start-date-id"
          endDate={this.props.filters.endDate}
          endDateId="end-date-id"
          onDatesChange={this.onDatesChange}
          focusedInput={this.state.calendarFocused}
          onFocusChange={this.onCalendarFocusChange}
          showClearDates={true}
          numberOfMonths={1}
          isOutsideRange={() => false}
        />
    
      </div>
    );
  }
}

const mapStateToProps = (state) => ({
  filters: state.filters
})

const mapDispatchToProps = (dispatch) => ({
  setTextFilter: ({ text }) => dispatch(iltersActions.setTextFilter({ text })),
  setSortByDate: () => dispatch(filtersActions.sortByDate()),
  setSortByAmount: () => dispatch(filtersActions.sortByAmount()),
  setStartDate: ({ startDate }) => dispatch(filtersActions.setStartDate({ startDate })),
  setEndDate: ({ endDate }) => dispatch(filtersActions.setEndDate({ endDate }))
})

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);