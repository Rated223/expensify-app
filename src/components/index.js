import AddExpensePage from './AddExpensePage';
import EditExpensePage from './EditExpensePage';
import ExpenseDashboardPage from './ExpenseDashboardPage';
import ExpenseList from './ExpenseList';
import ExpenseListItem from './ExpenseListItem';
import ExpenseListFilters from "./ExpenseListFilters";
import Header from './Header';
import NotFoundPage from './NotFoundPage';
import ExpenseForm from './ExpenseForm';
import ExpensesTotal from './ExpensesTotal';
import LoginPage from './LoginPage';
import LoadingPage from './LoadingPage'

export {
  AddExpensePage,
  EditExpensePage,
  ExpenseDashboardPage,
  ExpenseList,
  ExpenseListItem,
  ExpenseListFilters,
  Header,
  NotFoundPage,
  ExpenseForm,
  ExpensesTotal,
  LoginPage,
  LoadingPage
}