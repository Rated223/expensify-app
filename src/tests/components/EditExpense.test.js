import React from 'react';
import { shallow } from 'enzyme';
import { EditExpensePage } from '../../components/EditExpensePage';
import expenses from '../fixtures/expenses';
import { ExpenseForm } from '../../components';

let editExpense, startRemoveExpense, history, wrapper;

beforeEach(() => {
  editExpense = jest.fn();
  startRemoveExpense = jest.fn();
  history = { push: jest.fn() };
  wrapper = shallow(<EditExpensePage editExpense={editExpense} startRemoveExpense={startRemoveExpense} history={history} expense={expenses[0]}/>)
})

test('Should render Editpage correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

test('Should edit the expense', () => {
  const newExpense = expenses[0];
  newExpense.description = "Newwww description";
  wrapper.find('ExpenseForm').prop('onSubmit')(newExpense);
  expect(history.push).toHaveBeenLastCalledWith('/');
  expect(editExpense).toHaveBeenLastCalledWith({ id: newExpense.id, expense: newExpense });
});

test('Should remove the expense', () => {
  wrapper.find('button').prop('onClick')();
  expect(history.push).toHaveBeenLastCalledWith('/');
  expect(startRemoveExpense).toHaveBeenLastCalledWith({ id: expenses[0].id });
})