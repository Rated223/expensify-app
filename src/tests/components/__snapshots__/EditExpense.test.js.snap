// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render Editpage correctly 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title"
      >
        Edit expense
      </h1>
    </div>
  </div>
  <div
    className="content-container"
  >
    <div
      className="content-container__left"
    >
      <button
        className="button--secondary"
        onClick={[Function]}
      >
        Remove expense
      </button>
    </div>
    <ExpenseForm
      expense={
        Object {
          "amount": 195,
          "createdAt": 0,
          "description": "gum",
          "id": "1",
          "note": "",
        }
      }
      onSubmit={[Function]}
    />
  </div>
</div>
`;
